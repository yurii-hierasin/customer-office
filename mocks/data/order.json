{
  "id": 19948,
  "invoice": "US288669",
  "user_id": "224611",
  "status": "paid",
  "created": "2020-11-01T14:20:20.911+00:00",
  "updated": "2020-11-01T14:20:23.055+00:00",
  "total": 554,
  "service_list": [
    {
      "id": 21855,
      "order_id": 19948,
      "currency": {
        "code": "USD",
        "numeric": "840",
        "label": "US Dollar"
      },
      "total": 554,
      "created": "2020-11-01T14:20:20.923+00:00",
      "updated": "2020-11-01T14:20:20.923+00:00",
      "status": "pending",
      "cost_list": {
        "keys": {
          "proc_time": {
            "id": 50,
            "name": "12-15 business days",
            "days_count": 15
          },
          "visa_type": {
            "id": 19,
            "name": "Multiple entry"
          },
          "docs_count": null,
          "validity": null
        },
        "fields": [
          {
            "id": "0",
            "value": 345,
            "label": "Embassy fee",
            "currency": {
              "code": "USD",
              "numeric": "840",
              "label": "US Dollar"
            },
            "fieldCost": 345,
            "fieldTax": 0,
            "total": 345,
            "taxed": false,
            "zero_hidden": false
          },
          {
            "id": "1",
            "value": 209,
            "label": "Service fee",
            "currency": {
              "code": "USD",
              "numeric": "840",
              "label": "US Dollar"
            },
            "fieldCost": 209,
            "fieldTax": 0,
            "total": 209,
            "taxed": false,
            "zero_hidden": false
          }
        ]
      },
      "service": {
        "id": 5,
        "label": "Visa Service",
        "keys": [
          "residency",
          "citizenship",
          "destination",
          "jurisdiction",
          "visa_group"
        ],
        "prices": {
          "keys": [
            "visa_type",
            "proc_time"
          ],
          "fields": [
            {
              "id": 0,
              "label": "Embassy fee",
              "is_taxed": false,
              "zero_hidden": false
            },
            {
              "id": 1,
              "label": "Service fee",
              "is_taxed": true,
              "zero_hidden": false
            }
          ]
        },
        "identifier": "visa"
      },
      "service_item": {
        "id": 22794860,
        "service_id": 5,
        "residency": "US",
        "citizenship": "US",
        "destination": "UA",
        "visa_group": 220,
        "jurisdiction": 1,
        "prices": [
          {
            "keys": {
              "proc_time": {
                "id": 50,
                "name": "12-15 business days",
                "days_count": 15
              },
              "visa_type": {
                "id": 19,
                "name": "Multiple entry"
              },
              "docs_count": null,
              "validity": null
            },
            "fields": [
              {
                "id": "0",
                "value": 345,
                "label": "Embassy fee",
                "currency": null,
                "fieldCost": 345,
                "fieldTax": 0,
                "total": 345,
                "taxed": false,
                "zero_hidden": false
              },
              {
                "id": "1",
                "value": 209,
                "label": "Service fee",
                "currency": null,
                "fieldCost": 209,
                "fieldTax": 0,
                "total": 209,
                "taxed": false,
                "zero_hidden": false
              }
            ]
          }
        ],
        "offline": false
      },
      "key_questions": {
        "sex": [
          "Male"
        ],
        "email": [
          "john.doe@example.com"
        ],
        "mrstatus": [
          null
        ],
        "datebirth": [
          "1987-09-15"
        ],
        "entrydate": [
          null
        ],
        "last-name": [
          "Doe"
        ],
        "pspexdate": [
          null
        ],
        "emplstatus": [
          null
        ],
        "first-name": [
          "John"
        ],
        "placebirth": [
          null
        ],
        "previouscitizenship": [
          null
        ]
      },
      "meta_info": {},
      "tax_rate": 0,
      "conclusion_url": null
    }
  ],
  "shipping_list": [
    {
      "id": 737,
      "order_id": 19948,
      "currency": {
        "code": "USD",
        "numeric": "840",
        "label": "US Dollar"
      },
      "status": "completed",
      "total": 0,
      "from_client": false,
      "sender": null,
      "recipient": null,
      "type": "local",
      "shipment_name": "Local pickup",
      "default_prices": true,
      "service_name": "shipping.local",
      "additional_options": null
    }
  ],
  "source": null,
  "invoice_url": null,
  "affiliate": null
}
