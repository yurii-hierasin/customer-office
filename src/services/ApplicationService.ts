import {IDestination, IServiceListItem, IShipment, IVisaGroup, ShipmentTypesEnum} from '../store/retail/interfaces';

export default class ApplicationService {

    constructor(
        private application: IServiceListItem,
    ) {
    }

    getApplication(): IServiceListItem {
        return this.application
    }

    getCreatedDate(locale: string): string {
        const options = { year: 'numeric', month: 'short', day: 'numeric' }

        return new Intl.DateTimeFormat(locale, options).format(new Date(this.application.created))
    }

    getServiceDescription(
        visaGroups: IVisaGroup[],
        destinations: IDestination[],
    ): string {
        const description = []

        if (this.application.cost_list.keys.visa_type) {
            description.push(this.application.cost_list.keys.visa_type.name)

            const visaGroup = visaGroups.find(group => group.id === this.application.service_item.visa_group)
            visaGroup && description.push(visaGroup.label)
        }

        if (this.application.service.keys.includes('destination')) {
            const destination = destinations.find(dest => dest.alpha2 === this.application.service_item.destination)
            destination && description.push(destination.label)
        }

        return description.join(' / ')
    }

    getFullName(): string {
        return [
            this.application.key_questions['first-name'] || '',
            this.application.key_questions['last-name'] || '',
        ].join(' ')
    }

    getTotal(): string {
        return `${this.application.total.toFixed(2)} ${this.application.currency.code}`
    }

    getShippingLabel(shippingType: ShipmentTypesEnum, shippingList: IShipment[]) {
        let label = 'None'

        const shipping = shippingList.find((shipping) => {
            return shipping.type === shippingType
        })

        if (shipping) {
            label = shipping.shipment_name
        }

        return label
    }

}
