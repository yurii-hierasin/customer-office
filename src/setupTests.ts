// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom/extend-expect';
import {configure, ReactWrapper, ShallowWrapper} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import configureStore from './store';
import RetailAPI from './api/retailAPI';
import AuthAPI from './api/authAPI';

configure({ adapter: new Adapter() })

export const findByTestAttr = (component: ReactWrapper | ShallowWrapper, attr: string) => {
    return component.find(`[data-test='${attr}']`);
}

export const getStore = () => {
    return configureStore({retail: new RetailAPI(), auth: new AuthAPI()})
}
