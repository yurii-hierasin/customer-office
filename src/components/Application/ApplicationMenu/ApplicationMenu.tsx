import React, {useState} from 'react';
import uploadIcon from '../../../icons/Upload.svg';
import editIcon from '../../../icons/Edit.svg';
import {Dropdown} from 'react-bootstrap';
import {IOrder, ShipmentTypesEnum} from '../../../store/retail/interfaces';
import ApplicationService from '../../../services/ApplicationService';
import DocsModal from '../../docs-modal/DocsModal';
import ApplicationEditModal from '../../ApplicationEditModal/ApplicationEditModal';
import TreeDotsToggle from '../../TreeDotsToggle';
import DocumentsService from '../../../services/documents.service';


interface IApplicationMenuProps {
    order: IOrder,
    applicationService: ApplicationService
}

const ApplicationMenu = ({order, applicationService}: IApplicationMenuProps) => {
    const [showDocsModal, setDocsModalShow] = useState(false);
    const [showEditModal, setEditModalShow] = useState(false);

    const handleFilesModalClose = () => setDocsModalShow(false);
    const handleFilesModalShow = () => setDocsModalShow(true);

    const handleApplicationEditModalClose = () => setEditModalShow(false);
    const handleApplicationEditModalShow = () => setEditModalShow(true);

    return (
        <Dropdown>
            <Dropdown.Toggle as={TreeDotsToggle}/>
            <Dropdown.Menu>
                <Dropdown.Item
                    as="button"
                    onClick={handleApplicationEditModalShow}
                    data-test="edit-btn"
                >
                    <img src={editIcon} alt="Edit application"/>
                    <span className="pl-1" data-test="edit">Edit</span>
                </Dropdown.Item>
                <Dropdown.Item
                    as="button"
                    onClick={handleFilesModalShow}
                    data-test="file-btn"
                >
                    <img src={uploadIcon} alt="Upload|Download files"/>
                    <span className="pl-1" data-test="file">File</span>
                </Dropdown.Item>
                <Dropdown.Divider/>
                <Dropdown.ItemText>
                    <p>Service details:</p>
                    <p>
                        <span>Order: </span>
                        <span data-test="order-number">{order.invoice}</span>
                    </p>
                    <p>
                        <span>Inbound shipping: </span>
                        <span data-test="shipping-inbound-label">
                        {applicationService.getShippingLabel(ShipmentTypesEnum.inbound, order.shipping_list)}
                    </span>
                    </p>
                    <p>
                        <span>Return shipping: </span>
                        <span data-test="shipping-return-label">
                        {applicationService.getShippingLabel(ShipmentTypesEnum.outbound, order.shipping_list)}
                    </span>
                    </p>
                </Dropdown.ItemText>

                {showDocsModal && (
                    <DocsModal
                        order={order}
                        orderServiceItem={applicationService.getApplication()}
                        handleClose={handleFilesModalClose}
                        documentsService={new DocumentsService()}
                    />
                )}

                {showEditModal && <ApplicationEditModal
                    show={showEditModal}
                    handleClose={handleApplicationEditModalClose}
                />}
            </Dropdown.Menu>
        </Dropdown>
    )
}

export default ApplicationMenu
