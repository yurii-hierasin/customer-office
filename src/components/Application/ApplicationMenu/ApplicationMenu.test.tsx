import React from 'react';
import {shallow, ShallowWrapper} from 'enzyme';
import ApplicationMenu from './ApplicationMenu';
import order from '../../../../mocks/data/order.json'
import ApplicationService from '../../../services/ApplicationService';
import {findByTestAttr} from '../../../setupTests';

const setUp = () => {
    return shallow(<ApplicationMenu
        order={order}
        applicationService={(new ApplicationService(order.service_list[0]))}
    />)
}

describe('ApplicationMenu component', () => {

    let wrapper: ShallowWrapper

    beforeEach(() => wrapper = setUp())

    test('should render', () => {
        expect(wrapper).toMatchSnapshot()
    })

    test('should get order number', () => {
        expect(findByTestAttr(wrapper, 'order-number').text())
            .toBe('US288669')
    })

    test('should be inbound shipping default label', () => {
        expect(findByTestAttr(wrapper, 'shipping-inbound-label').text())
            .toBe('None')
    })

    test('should be return shipping default label', () => {
        expect(findByTestAttr(wrapper, 'shipping-return-label').text())
            .toBe('None')
    })

    test('file management component appear after onClick', () => {
        expect(wrapper.find('DocsModal').length).toBe(0)
        findByTestAttr(wrapper, 'file-btn').simulate('click')
        expect(wrapper.find('DocsModal').length).toBe(1)
    })

    test('edit component appear after onClick', () => {
        expect(wrapper.find('ApplicationEditModal').length).toBe(0)
        findByTestAttr(wrapper, 'edit-btn').simulate('click')
        expect(wrapper.find('ApplicationEditModal').length).toBe(1)
    })

})
