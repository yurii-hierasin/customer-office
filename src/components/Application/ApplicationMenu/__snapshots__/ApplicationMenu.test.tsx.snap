// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationMenu component should render 1`] = `
<DropdownButton
  drop="left"
  title={
    <img
      alt="menu"
      className="mb-1"
      src="TreeDotsVertical.svg"
    />
  }
>
  <DropdownItem
    as="button"
    disabled={false}
  >
    <img
      alt="Edit application"
      src="Edit.svg"
    />
    <span
      className="pl-1"
      data-test="edit"
    >
      Edit
    </span>
  </DropdownItem>
  <DropdownItem
    as="button"
    disabled={false}
  >
    <img
      alt="Upload|Download files"
      src="Upload.svg"
    />
    <span
      className="pl-1"
      data-test="file"
    >
      File
    </span>
  </DropdownItem>
  <DropdownDivider
    role="separator"
  />
  <DropdownItemText>
    <p>
      Service details:
    </p>
    <p>
      <span>
        Order: 
      </span>
      <span
        data-test="order-number"
      >
        US288669
      </span>
    </p>
    <p>
      <span>
        Inbound shipping: 
      </span>
      <span
        data-test="shipping-inbound-label"
      >
        None
      </span>
    </p>
    <p>
      <span>
        Return shipping: 
      </span>
      <span
        data-test="shipping-return-label"
      >
        None
      </span>
    </p>
  </DropdownItemText>
</DropdownButton>
`;
